### PR TITLE
General revisioning improvements

### PR DESCRIPTION
**Pull request by a-drew**
## Summary
General improvements to the package

- test:  Significantly more coverage, covering different relation types and more complex scenarios
- refactor: split up unwatched and excluded columns as intended originally
- refactor: deleting/restoring/revisioning children will now always perform a full copy when duplicating your model
- refactor: cascade delete the revision when deleting a revisionable model
- feat: preserve linear history when deleting revisions
- feat: register custom handlers for relations revisioning, by name or type
- feat: add a relation to retrieve newest item in revision lineage

### CODE REVIEW

File: src/Models/Revision.php [hunk 1]
```diff
@@ -115,6 +115,22 @@ class Revision extends MorphPivot
...
+            if ($next !== null) {
+                $next->previous_revision_id = $revision->previous_revision_id;
+                $next->save();
+            } elseif ($revision->previous_revision_id !== null) {
+                // if no newer revision exists, update the latest flag on the previous revision
+                $revision->previous()->update(['latest' => true]);
+            }
+
+        });
+    }
```
[the remark this answers is not in the document]
Comment (kfriars) [reply src/Models/Revision.php:132]: Right... These are revisions, not revisionables. 

And for the 2nd one you are right, I didn't take the inital condition into consideration, that guard is technically `$next === null && $revision->previous_revision_id !== null`.

File: src/Concerns/HasRevisions.php [hunk 2]
```diff
@@ -363,17 +357,77 @@ public function performRevision()
...
+                $this->{'revision' . ucfirst($relation)}($copy, $relation, $type); // revisionModules()
+            } elseif (method_exists($this, 'revision' . substr($type, strrpos($type, '\\') + 1))) {
+                $this->{'revision' . ucfirst($type)}($copy, $relation, $type); // revisionHasMany()
+            } elseif ($relationHelper::isChild($type)) {
+                $this->revisionChildren($copy, $relation);
+            } elseif ($relationHelper::isPivoted($type)) { // default is pivot
+                $changes = $copy->$relation()->syncWithoutDetaching($this->$relation()->get());
+            } else {
+                // skipped relation
+            }
```
[the remark this answers is not in the document]
Comment (kfriars) [reply src/Concerns/HasRevisions.php:402]: Right. I like the convention, but `revisionChildren()` could easily cause issues. Maybe better to make it something more verbose, like `method_exists($this, 'revision' . ucfirst($relation).'Relation')` which will make it highly unlikely to cause an issue.

File: src/Helpers/RelationHelper.php [hunk 3]
```diff
@@ -211,4 +212,47 @@ public static function getModelRelations(Model $model, $refresh = false): array
...
+    {
+        if ($merge) {
+            static::$relationTypes = array_merge(static::$relationTypes, $types);
+        } else {
+            static::$relationTypes = array_merge([
+                'hasOne', 'hasMany', 'hasManyThrough', 'belongsTo', 'belongsToMany',
+                'morphOne', 'morphMany', 'morphTo', 'morphToMany',
+            ], $types);
+        }
+    }
```
[the remark this answers is not in the document]
Comment (a-drew) [reply src/Helpers/RelationHelper.php]: split up and added a test

File: src/Concerns/HasRevisions.php (hunk 2 above)
[the remark this answers is not in the document]
Comment (a-drew) [reply src/Concerns/HasRevisions.php]: Changes made:
- renamed `revisionRelations()` to `replicateRelationsTo($copy)`, following the logic that revisioning isn't guaranteed to occur
- renamed `revisionChildren()` to `replicateChildrenTo($copy, $relation)`
- custom handlers have to be registered as `replicate{Relation}Relation()` or `replicate{RelationType}Relations()`

**Pull request merged**